{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "__doc__ = \"\"\"Calculate z-normalized tetranucleotide frequency from a FASTA file.\n",
    "\n",
    "Input: A fasta file containing only A, C, G, T and N sequences\n",
    "Output: A (n_sequences * 136) matrix of tetranucleotide frequencies,\n",
    "z-normalized through axis 0.\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys as _sys\n",
    "import os as _os\n",
    "import argparse as _argparse\n",
    "import numpy as _np\n",
    "import gzip as _gzip\n",
    "\n",
    "if __package__ is None or __package__ == '':\n",
    "    import vambtools as _vambtools\n",
    "    \n",
    "else:\n",
    "    import vamb.vambtools as _vambtools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "TNF_HEADER = '#contigheader\\t' + '\\t'.join([\n",
    "'AAAA/TTTT', 'AAAC/GTTT', 'AAAG/CTTT', 'AAAT/ATTT',\n",
    "'AACA/TGTT', 'AACC/GGTT', 'AACG/CGTT', 'AACT/AGTT',\n",
    "'AAGA/TCTT', 'AAGC/GCTT', 'AAGG/CCTT', 'AAGT/ACTT',\n",
    "'AATA/TATT', 'AATC/GATT', 'AATG/CATT', 'AATT',\n",
    "'ACAA/TTGT', 'ACAC/GTGT', 'ACAG/CTGT', 'ACAT/ATGT',\n",
    "'ACCA/TGGT', 'ACCC/GGGT', 'ACCG/CGGT', 'ACCT/AGGT',\n",
    "'ACGA/TCGT', 'ACGC/GCGT', 'ACGG/CCGT', 'ACGT',\n",
    "'ACTA/TAGT', 'ACTC/GAGT', 'ACTG/CAGT', 'AGAA/TTCT',\n",
    "'AGAC/GTCT', 'AGAG/CTCT', 'AGAT/ATCT', 'AGCA/TGCT',\n",
    "'AGCC/GGCT', 'AGCG/CGCT', 'AGCT',      'AGGA/TCCT',\n",
    "'AGGC/GCCT', 'AGGG/CCCT', 'AGTA/TACT', 'AGTC/GACT',\n",
    "'AGTG/CACT', 'ATAA/TTAT', 'ATAC/GTAT', 'ATAG/CTAT',\n",
    "'ATAT',      'ATCA/TGAT', 'ATCC/GGAT', 'ATCG/CGAT',\n",
    "'ATGA/TCAT', 'ATGC/GCAT', 'ATGG/CCAT', 'ATTA/TAAT',\n",
    "'ATTC/GAAT', 'ATTG/CAAT', 'CAAA/TTTG', 'CAAC/GTTG',\n",
    "'CAAG/CTTG', 'CACA/TGTG', 'CACC/GGTG', 'CACG/CGTG',\n",
    "'CAGA/TCTG', 'CAGC/GCTG', 'CAGG/CCTG', 'CATA/TATG',\n",
    "'CATC/GATG', 'CATG',      'CCAA/TTGG', 'CCAC/GTGG',\n",
    "'CCAG/CTGG', 'CCCA/TGGG', 'CCCC/GGGG', 'CCCG/CGGG',\n",
    "'CCGA/TCGG', 'CCGC/GCGG', 'CCGG',      'CCTA/TAGG',\n",
    "'CCTC/GAGG', 'CGAA/TTCG', 'CGAC/GTCG', 'CGAG/CTCG',\n",
    "'CGCA/TGCG', 'CGCC/GGCG', 'CGCG',      'CGGA/TCCG',\n",
    "'CGGC/GCCG', 'CGTA/TACG', 'CGTC/GACG', 'CTAA/TTAG',\n",
    "'CTAC/GTAG', 'CTAG',      'CTCA/TGAG', 'CTCC/GGAG',\n",
    "'CTGA/TCAG', 'CTGC/GCAG', 'CTTA/TAAG', 'CTTC/GAAG',\n",
    "'GAAA/TTTC', 'GAAC/GTTC', 'GACA/TGTC', 'GACC/GGTC',\n",
    "'GAGA/TCTC', 'GAGC/GCTC', 'GATA/TATC', 'GATC',\n",
    "'GCAA/TTGC', 'GCAC/GTGC', 'GCCA/TGGC', 'GCCC/GGGC',\n",
    "'GCGA/TCGC', 'GCGC',      'GCTA/TAGC', 'GGAA/TTCC',\n",
    "'GGAC/GTCC', 'GGCA/TGCC', 'GGCC', 'GGGA/TCCC',\n",
    "'GGTA/TACC', 'GTAA/TTAC', 'GTAC', 'GTCA/TGAC',\n",
    "'GTGA/TCAC', 'GTTA/TAAC', 'TAAA/TTTA', 'TACA/TGTA',\n",
    "'TAGA/TCTA', 'TATA',      'TCAA/TTGA', 'TCCA/TGGA',\n",
    "'TCGA',      'TGAA/TTCA', 'TGCA', 'TTAA']) + '\\n'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_contigs(contigpath, minlength=2000):\n",
    "    \"\"\"Parses a FASTA file and produces a list of headers and a matrix of TNF.\n",
    "    \n",
    "    Input:\n",
    "        contigpath: Path to a FASTA file with contigs\n",
    "        min_length[2000]: Minimum length of contigs\n",
    "    \n",
    "    Outputs:\n",
    "        contignames: A list of contig headers\n",
    "        tnfs: A (n_FASTA_entries x 136) matrix of tetranucleotide freq.\n",
    "    \"\"\"\n",
    "    \n",
    "    tnf_list = list()\n",
    "    contignames = list()\n",
    "\n",
    "    with _vambtools.Reader(contigpath, 'rb') as contigfile:\n",
    "        entries = _vambtools.byte_iterfasta(contigfile)\n",
    "\n",
    "        for entry in entries:\n",
    "            if len(entry) < minlength:\n",
    "                continue\n",
    "                \n",
    "            tnf_list.append(entry.fourmer_freq())\n",
    "            contignames.append(entry.header)\n",
    "            \n",
    "    tnfs = _np.array(tnf_list, dtype=_np.float32)\n",
    "    del tnf_list\n",
    "    \n",
    "    _vambtools.zscore(tnfs, axis=0, inplace=True)\n",
    "    \n",
    "    return tnfs, contignames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_tnf(tnfpath, contignames, tnfs):\n",
    "    \"\"\"Writes a TNF table to specified output path.\n",
    "    \n",
    "    Input:\n",
    "        tnfpath: Path to write output\n",
    "        contignames: A list of contig headers\n",
    "        tnfs: A (n_FASTA_entries x 136) matrix of tetranucleotide freq.\n",
    "    \n",
    "    Outputs: None\n",
    "    \"\"\"\n",
    "    \n",
    "    formatstring = '{}\\t' + '\\t'.join(['{:.4f}']*136) + '\\n'\n",
    "    \n",
    "    with _gzip.open(tnfpath, 'w') as file:\n",
    "        file.write(TNF_HEADER.encode())\n",
    "        for contigname, tnf in zip(contignames, tnfs):\n",
    "            file.write(formatstring.format(contigname, *tnf).encode())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == '__main__':\n",
    "    parserkws = {'prog': 'calculate_tnf.py',\n",
    "                 'formatter_class': _argparse.RawDescriptionHelpFormatter,\n",
    "                 'usage': 'parsecontigs.py contigs.fna(.gz) tnfout lengthsout',\n",
    "                 'description': __doc__}\n",
    "\n",
    "    # Create the parser\n",
    "    parser = _argparse.ArgumentParser(**parserkws)\n",
    "\n",
    "    parser.add_argument('contigs', help='FASTA file of contigs')\n",
    "    parser.add_argument('tnfout', help='TNF output path')\n",
    "    \n",
    "    parser.add_argument('-m', dest='minlength', type=int, default=2000,\n",
    "                        help='minimum length of contigs [2000]')\n",
    "\n",
    "    # Print help if no arguments are given\n",
    "    if len(_sys.argv) == 1:\n",
    "        parser.print_help()\n",
    "        _sys.exit()\n",
    "\n",
    "    args = parser.parse_args()\n",
    "    \n",
    "    if args.minlength < 4:\n",
    "        raise ValueError('Minlength must be at least 4')\n",
    "\n",
    "    if not _os.path.isfile(args.contigs):\n",
    "        raise FileNotFoundError(args.contigs)\n",
    "\n",
    "    if _os.path.exists(args.tnfout):\n",
    "        raise FileExistsError(args.tnfout)\n",
    "    \n",
    "    contignames, tnfs = read_contigs(args.contigs, args.minlength)\n",
    "    write_tnf(args.tnfout, contignames, tnfs)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
