{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Metagenomic RPKM estimator\n",
    "\n",
    "# Author: Jakob Nybo Nissen, DTU Bioinformatics, jakni@bioinformatics.dtu.dk\n",
    "# Date: 2018-04-24\n",
    "    \n",
    "# This script calculates RPKM, when paired end reads are mapped to a contig\n",
    "# catalogue with BWA MEM. It will not be accurate with single end reads or\n",
    "# any other mapper than BWA MEM.\n",
    "\n",
    "# Theory:\n",
    "# We want a simple way to estimate abundance of redundant contig catalogues.\n",
    "# Earlier we used to run analysis on deduplicated gene catalogues, but since \n",
    "# both depth and kmer composition are only stable for longer contigs, we have\n",
    "# moved to contig catalogues. We have not found a way of deduplicating contigs.\n",
    "\n",
    "# For this we have until now used two methods:\n",
    "# 1) Only counting the primary hits. In this case the read will never be\n",
    "# assigned to any contig which differ by just 1 basepair. Even for\n",
    "# identical contigs, reads are assigned randomly which causes noise.\n",
    "\n",
    "# 2) Using MetaBAT's jgi_summarize_bam_contig_depths, a script which is not\n",
    "# documented and we cannot figure out how works. When testing with small\n",
    "# toy data, it produces absurd results.\n",
    "\n",
    "# This script is an attempt to take an approach as simple as possible while\n",
    "# still being sound technically. We simply count the number of reads in a\n",
    "# contig normalized by contig length and total number of reads.\n",
    "\n",
    "# We look at all hits, including secondary hits. We do not discount partial\n",
    "# alignments. Also, if a read maps to multiple contigs, we don't count each hit\n",
    "# as less than if it mapped to both. The reason for all these decisions is that\n",
    "# if the aligner believes it's a hit, we believe the contig is present.\n",
    "\n",
    "# We do not take varying insert sizes into account. It is unlikely that\n",
    "# any contig with enough reads to provide a reliable estimate of depth would,\n",
    "# by chance, only recruit read pairs with short or long insert size. So this\n",
    "# will average out over all contigs.\n",
    "\n",
    "# We DO filter the input file for duplicate lines, as BWA MEM erroneously\n",
    "# produces quite a few of them.\n",
    "\n",
    "# We count each read independently, because BWA MEM often assigns mating reads\n",
    "# to different contigs. If a read has their mate unmapped, we count it twice\n",
    "# to compensate (one single read corresponds to two paired reads)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "__doc__ = \"\"\"Metagenomic RPKM estimator\n",
    "\n",
    "Author: Jakob Nybo Nissen, DTU Bioinformatics, jakni@bioinformatics.dtu.dk\n",
    "    \n",
    "This script calculates reads per kilobase per million mapped reads (RPKM),\n",
    "when paired end reads are mapped to a contig catalogue with BWA MEM.\n",
    "Because of the idiosyncracities of BWA, it will not be accurate with single end\n",
    "reads or any other mapper than BWA MEM.\n",
    "\n",
    "Requires the module pysam to run.\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam as _pysam\n",
    "\n",
    "import sys as _sys\n",
    "import os as _os\n",
    "import multiprocessing as _multiprocessing\n",
    "import numpy as _np\n",
    "import gzip as _gzip\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    import argparse as _argparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "DEFAULT_PROCESSES = min(8, _os.cpu_count())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _get_all_references(alignedsegment):\n",
    "    \"\"\"Given a pysam aligned segment, returns a list with the names of all\n",
    "    references the read maps to, both primary and secondary hits.\n",
    "    \"\"\"\n",
    "    \n",
    "    references = [alignedsegment.reference_name]\n",
    "    \n",
    "    # Some reads don't have secondary hits\n",
    "    if not alignedsegment.has_tag('XA'):\n",
    "        return references\n",
    "    \n",
    "    # XA is a string contigname1,<other info>;contigname2,<other info>; ...\n",
    "    secondary_alignment_string = alignedsegment.get_tag('XA')\n",
    "    secondary_alignments = secondary_alignment_string.split(';')[:-1]\n",
    "    \n",
    "    for secondary_alignment in secondary_alignments:\n",
    "        references.append(secondary_alignment.partition(',')[0])\n",
    "        \n",
    "    return references"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _filter_segments(segmentiterator, minscore):\n",
    "    \"\"\"Returns an iterator of AlignedSegment filtered for reads with low\n",
    "    alignment score, and for any segments identical to the previous segment.\n",
    "    This is necessary as BWA MEM produces dopplegangers.\n",
    "    \"\"\"\n",
    "    \n",
    "    # First get the first segment, so we in the loop can compare to the previous\n",
    "    alignedsegment = next(segmentiterator)\n",
    "    \n",
    "    yield alignedsegment\n",
    "        \n",
    "    lastname = alignedsegment.query_name\n",
    "    lastwasforward = alignedsegment.flag & 64 == 64\n",
    "        \n",
    "    for alignedsegment in segmentiterator:\n",
    "        if alignedsegment.get_tag('AS') < minscore:\n",
    "            continue\n",
    "        \n",
    "        # Depressingly, BWA somtimes outputs the same read multiple times.\n",
    "        # We identify them by having same name as directionality as previous. \n",
    "        thisname = alignedsegment.query_name\n",
    "        thisisforward = alignedsegment.flag & 64 == 64\n",
    "        \n",
    "        if thisisforward is not lastwasforward or thisname != lastname:\n",
    "            yield alignedsegment\n",
    "            \n",
    "            lastname = thisname\n",
    "            lastwasforward = thisisforward"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _get_contig_rpkms(path, minscore=50, minlength=2000):\n",
    "    \"\"\"Returns  RPKM (reads per kilobase per million mapped reads)\n",
    "    for all contigs present in BAM header.\n",
    "    \n",
    "    Inputs:\n",
    "        path: Path to BAM file\n",
    "        minscore [50]: Minimum alignment score (AS field) to consider\n",
    "        minlength [2000]: Discard any references shorter than N bases \n",
    "        \n",
    "    Outputs:\n",
    "        path: Same as input path\n",
    "        rpkms A float32-array with RPKM for each contig in BAM header\n",
    "    \"\"\"\n",
    "    \n",
    "    bamfile = _pysam.AlignmentFile(path, \"rb\")\n",
    "\n",
    "    # We can only get secondary alignment reference names, not indices. So we must\n",
    "    # make an idof dict to look up the indices.\n",
    "    idof = {contig: i for i, contig in enumerate(bamfile.references)}\n",
    "    contiglengths = bamfile.lengths\n",
    "    halfreads = _np.zeros(len(contiglengths), dtype=_np.int32)\n",
    "    \n",
    "    nhalfreads = 0\n",
    "    for segment in _filter_segments(bamfile, minscore):\n",
    "        nhalfreads += 1\n",
    "        \n",
    "        # Read w. unmapped mates count twice as they represent a whole read\n",
    "        value = 2 if segment.mate_is_unmapped else 1\n",
    "        \n",
    "        for reference in _get_all_references(segment):\n",
    "            id = idof[reference]\n",
    "            halfreads[id] += value\n",
    "            \n",
    "    bamfile.close()\n",
    "    del idof\n",
    "    \n",
    "    rpkms = _np.zeros(len(contiglengths), dtype=_np.float32)\n",
    "    \n",
    "    millionmappedreads = nhalfreads / 1e6\n",
    "    \n",
    "    for i, (contiglength, nhalfreads) in enumerate(zip(contiglengths, halfreads)):\n",
    "        kilobases = contiglength / 1000\n",
    "        rpkms[i] = nhalfreads / (kilobases * millionmappedreads)\n",
    "        \n",
    "    # Now filter for small contigs\n",
    "    lengthmask = _np.array(contiglengths, dtype=_np.int32) >= minlength\n",
    "    rpkms = rpkms[lengthmask]\n",
    "    \n",
    "    return path, rpkms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_bamfiles(paths, minscore=50, minlength=100,\n",
    "                  processes=DEFAULT_PROCESSES):\n",
    "    \"Placeholder docstring - replaced after this func definition\"\n",
    "    \n",
    "    # Get references and lengths from first BAM file.\n",
    "    # We need these to print them in the output.\n",
    "    # Might as well do it before spawning all those processes.\n",
    "    firstfile = _pysam.AlignmentFile(paths[0], \"rb\")\n",
    "    contignames = list()\n",
    "    \n",
    "    for name, length in zip(firstfile.references, firstfile.lengths):\n",
    "        if length >= minlength:\n",
    "            contignames.append(name)\n",
    "    \n",
    "    ncontigs = len(contignames)\n",
    "    \n",
    "    if ncontigs == 0:\n",
    "        raise ValueError('No headers in first bam file after filtering')\n",
    "        \n",
    "    del firstfile\n",
    "    \n",
    "    # Spawn independent processed to calculate RPKM for each of the BAM files\n",
    "    processresults = list()\n",
    "\n",
    "    # Queue all the processes\n",
    "    with _multiprocessing.Pool(processes=processes) as pool:\n",
    "        for path in paths:\n",
    "            arguments = (path, minscore, minlength)\n",
    "            processresults.append(pool.apply_async(_get_contig_rpkms, arguments))\n",
    "        \n",
    "        # For some reason, this is needed.\n",
    "        pool.close()\n",
    "        pool.join()\n",
    "            \n",
    "    sample_rpkms = dict()\n",
    "    \n",
    "    for processresult in processresults:\n",
    "        if processresult.successful():\n",
    "            path, rpkm = processresult.get()\n",
    "            sample_rpkms[path] = rpkm\n",
    "            \n",
    "        else:\n",
    "            return processresult\n",
    "            raise _multiprocessing.ProcessError\n",
    "            \n",
    "    # Check results have same length\n",
    "    wrong = [(path, len(rpkm)) for path, rpkm in sample_rpkms.items()\n",
    "             if len(rpkm) != ncontigs]\n",
    "    \n",
    "    if wrong:\n",
    "        message = ('Expected number of headers: ' +\n",
    "                   str(ncontigs) +\n",
    "                   '\\nfollowing samples have wrong number of headers:\\n' +\n",
    "                   '\\n'.join(p+'\\t'+str(i) for p, i in wrong))\n",
    "        raise ValueError(message)\n",
    "            \n",
    "    return sample_rpkms, contignames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "read_bamfiles.__doc__ = \"\"\"Spawns processes to parse BAM files and get contig rpkms.\n",
    "    \n",
    "Input:\n",
    "    path: Path to BAM file\n",
    "    minscore [50]: Minimum alignment score (AS field) to consider\n",
    "    minlength [100]: Ignore any references shorter than N bases \n",
    "    processes [{}]: Number of processes to spawn\n",
    "\n",
    "Outputs:\n",
    "    sample_rpkms: A {{path: Numpy-32-float-RPKM}} dictionary\n",
    "    contignames: A list of contignames from first BAM header\n",
    "\"\"\".format(DEFAULT_PROCESSES)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_samplerpkm(path, sample_rpkms):\n",
    "    \"\"\"Saves an RPKM dictionary to a path as .npz file.\n",
    "    \n",
    "    Inputs:\n",
    "        path: Path to save to.\n",
    "        sample_rpkms: A {path: Numpy-32-float-RPKM} dictionary\n",
    "        \n",
    "    Output: None\n",
    "    \"\"\"\n",
    "    \n",
    "    _np.savez_compressed(path, **sample_rpkms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_npz(path, columns):\n",
    "    \"\"\"Creates a (n-contigs x n-bamfiles) array from an npz object as created\n",
    "    by sample_rpkms_tonpz.\n",
    "    \n",
    "    Inputs: \n",
    "        path: Path to npz object\n",
    "        columns: Names of BAM file paths in npz object in correct order\n",
    "        \n",
    "    Output: A (n-contigs x n-bamfiles) array with each column being the corre-\n",
    "    sponding numpy array from the npz file, normalized so each row sums to 1\n",
    "    \"\"\"\n",
    "    \n",
    "    npz = _np.load(path)\n",
    "    \n",
    "    ncontigs = len(npz[columns[0]])\n",
    "    ncolumns = len(columns)\n",
    "    \n",
    "    arr = _np.empty((ncontigs, ncolumns))\n",
    "    \n",
    "    for i, filename in enumerate(columns):\n",
    "        column = npz[filename]\n",
    "        \n",
    "        if not len(column) == ncontigs:\n",
    "            errormessage = 'Expected {} contigs, but {} has {}'\n",
    "            raise ValueError(errormessage.format(ncontigs, filename, len(column)))\n",
    "        \n",
    "        arr[:, i] = column\n",
    "        \n",
    "    npz.close()\n",
    "        \n",
    "    # Normalize to rows sum to 1\n",
    "    rowsums = _np.sum(arr, axis=1)\n",
    "    rowsums[rowsums == 0] = 1\n",
    "    arr /= rowsums.reshape((-1, 1))\n",
    "    \n",
    "    return arr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def toarray(sample_rpkms, columns):\n",
    "    \"\"\"Creates a (n-contigs x n-bamfiles) array from a sample_rpkms\n",
    "    (expected to be a {path: Numpy-32-float-RPKM} dictionary)\n",
    "    \n",
    "    Inputs: \n",
    "        sample_rpkms: A {path: Numpy-32-float-RPKM} dictionary\n",
    "        columns: Names of BAM file paths in sample_rpkms object in correct order\n",
    "        \n",
    "    Output: A (n-contigs x n-bamfiles) array with each column being the corre-\n",
    "    sponding array from sample_rpkms, normalized so each row sums to 1\n",
    "    \"\"\"\n",
    "    \n",
    "    ncontigs = len(next(iter(sample_rpkms.values())))\n",
    "    \n",
    "    for filename, array in sample_rpkms.items():\n",
    "        if not len(array) == ncontigs:\n",
    "            errormessage = 'Expected {} contigs, but {} has {}'\n",
    "            raise ValueError(errormessage.format(ncontigs, filename, len(array)))\n",
    "    \n",
    "    arr = _np.zeros((ncontigs, len(sample_rpkms)), dtype=_np.float32)\n",
    "    \n",
    "    for n, path in enumerate(columns):\n",
    "        rpkm = sample_rpkms[path]\n",
    "        arr[:, n] = rpkm\n",
    "    \n",
    "    # Normalize to rows sum to 1\n",
    "    rowsums = _np.sum(arr, axis=1)\n",
    "    rowsums[rowsums == 0] = 1\n",
    "    arr /= rowsums.reshape((-1, 1))\n",
    "    \n",
    "    return arr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_rpkms(filehandle, sample_rpkms, columns):\n",
    "    \"\"\"Writes an RPKMs dict to specified output path.\n",
    "    \n",
    "    Input:\n",
    "        filehandle: Open filehandle to write to\n",
    "        sample_rpkms: A {path: Numpy-32-float-RPKM} dictionary\n",
    "        columns: Names of BAM file paths in sample_rpkms object in correct order\n",
    "    Outputs: None\n",
    "    \"\"\"\n",
    "    \n",
    "    rpkms = [sample_rpkms[column] for column in columns]\n",
    "    \n",
    "    formatstring = '\\t'.join(['{:.4f}']*len(rpkms)) + '\\n'\n",
    "    header = '# ' + '\\t'.join(columns) + '\\n'\n",
    "    \n",
    "    filehandle.write(header)\n",
    "    for values in zip(*rpkms):\n",
    "        filehandle.write(formatstring.format(*values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == '__main__':\n",
    "    usage = \"python parsebam.py [OPTION ...] OUTFILE BAMFILE [BAMFILE ...]\"\n",
    "    parser = _argparse.ArgumentParser(\n",
    "        description=__doc__,\n",
    "        formatter_class=_argparse.RawDescriptionHelpFormatter,\n",
    "        usage=usage)\n",
    "    \n",
    "    # Get the maximal number of extra processes to spawn\n",
    "    cpus = min(8, _os.cpu_count())\n",
    "    \n",
    "    # Positional arguments\n",
    "    parser.add_argument('outfile', help='path to output file')\n",
    "    parser.add_argument('bamfiles', help='relative path to find bam files', nargs='+')\n",
    "    \n",
    "    # Optional arguments\n",
    "    parser.add_argument('-a', dest='minscore', type=int, default=50,\n",
    "                        help='minimum alignment score [50]')\n",
    "    parser.add_argument('-m', dest='minlength', type=int, default=100,\n",
    "                        help='discard any references shorter than N [100]')\n",
    "    parser.add_argument('-p', dest='processes', type=int, default=cpus,\n",
    "                        help=('number of extra processes to spawn '\n",
    "                              '[min(' + str(cpus) + ', nfiles)]'))\n",
    "    \n",
    "    # If no arguments, print help\n",
    "    if len(_sys.argv) == 1:\n",
    "        parser.print_help()\n",
    "        _sys.exit()\n",
    "        \n",
    "    args = parser.parse_args()\n",
    "    \n",
    "    # Check number of cores\n",
    "    if args.processes < 1:\n",
    "        raise argsparse.ArgumentTypeError('Zero or negative processes requuested.')\n",
    "        \n",
    "    if args.minlength < 4:\n",
    "        raise ValueError('Minlength must be at least 4')\n",
    "        \n",
    "    # Check presence of output file\n",
    "    if _os.path.exists(args.outfile):\n",
    "        raise FileExistsError('Output file: ' + args.outfile)\n",
    "        \n",
    "    directory = _os.path.dirname(args.outfile)\n",
    "    if directory and not _os.path.isdir(directory):\n",
    "        raise NotADirectoryError(directory)\n",
    "        \n",
    "    # Check presence of all BAM files\n",
    "    for bamfile in args.bamfiles:\n",
    "        if not _os.path.isfile(bamfile):\n",
    "            raise FileNotFoundError('Bam file: ' + bamfile)\n",
    "            \n",
    "    # If more CPUs than files, scale down CPUs\n",
    "    if args.processes > len(args.bamfiles):\n",
    "        args.processes = len(args.bamfiles)\n",
    "        print('More processes given than files, scaling to {}.'.format(args.processes))\n",
    "    \n",
    "    print('Number of files:', len(args.bamfiles))\n",
    "    print('Number of parallel processes:', args.processes)\n",
    "    \n",
    "    sample_rpkms, contignames = read_bamfiles(args.bamfiles, args.minscore,\n",
    "                                              args.minlength, args.processes)\n",
    "    \n",
    "    with open(args.outfile, 'w') as file:\n",
    "        write_rpkms(file, sample_rpkms, args.bamfiles)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
